shoop/core/migrations/0013_address_changes.py

```diff
@@ -0,0 +1,166 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import models, migrations
+import django_countries.fields
+import shoop.core.utils.name_mixin
+import django.db.models.deletion
+import shoop.core.models._base
+import shoop.core.fields
+import shoop.utils.models
+
+
+def from_data(cls, data, immutable=False):
+    """
+    Create Address from a data dictionary.
+
+    :type data: dict[str,str]
+    :type immutable: boolean
+    :rtype: Unsaved Address
+    """
+    if immutable:
+        return cls.objects.get_or_create(**data)[0]
+    return cls(**data)
+
+
+def populate_immutable_addresses(cls, field, address_cls, address, address_data):
+    """
+    Update existing order address with immutable address which is got
+    Address from_data. This makes sure that only one identical immutable
+    address is created.
+
+    :param cls: Order class
+    :param field: field to update
+    :param address_cls: Address class
+    :param address: original address to be updated
+    :param address_data: address data
+    :return: None
+    """
+    query = {field: address}
+    objects = cls.objects.filter(**query)
+    if objects.exists():
+        immutable_address = from_data(address_cls, address_data, immutable=True)
+        immutable_address.save()
+        updated_field = {field: immutable_address}
+        objects.update(**updated_field)
+
+
+def copy_addresses(apps, schema_editor):
+    """
+    Copy existing addresses to new Mutable and Immutable addresses
+    For every address new MutableAddress is created with same id
+    For every address linked to order identical immutable address
+    is updated for certain order.
+    """
+    address_cls = apps.get_model("shoop", "Address")
+    order_cls = apps.get_model("shoop", "Order")
+    mutable_address_cls = apps.get_model("shoop", "MutableAddress")
+    immutable_address_cls = apps.get_model("shoop", "ImmutableAddress")
+    db_alias = schema_editor.connection.alias
+
+    for address in address_cls.objects.using(db_alias).all():
+        address_data_dict = shoop.utils.models.get_data_dict(address)
+        address_data_dict.pop("is_immutable")
+        mutable_address = from_data(mutable_address_cls, address_data_dict)
+        mutable_address.id = address.id
+        mutable_address.save()
+
+        populate_immutable_addresses(
+            order_cls, "billing_address", immutable_address_cls, address, address_data_dict)
+        populate_immutable_addresses(
+            order_cls, "shipping_address", immutable_address_cls, address, address_data_dict)
+
+
+class Migration(migrations.Migration):
```

Comment:
Could this migration maybe rename the old single address table to either mutable or immutable? I mean something like


* Create `ImmutableOrder` table
* Rename `Order.billing_address` to `Order.mutable_billing_address`
* Rename `Order.shipping_address` to `Order.mutable_shipping_address`
* Add `Order.billing_address` FK (to immutable table)
* Add `Order.shipping_address` FK (to immutable table)
* Copy mutable addresses to immutable addresses and backfill FKs in Order
* Remove `Address.immutable`
* Rename `Address` table to `MutableAddress`
* Remove `Order.mutable_*`